modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/GlobalSnapshotStateChannelManager.scala

```diff
@@ -0,0 +1,125 @@
+            }
+
+        private def acceptForAddress(ordinal: SnapshotOrdinal)(lastHash: Hash, outputs: List[StateChannelOutput]) = for {
+          outputsWithHashes <- outputs.traverse(stateChannelOutputWithHashes)
+          possibleCandidates = onlyPossibleReferences(lastHash, outputsWithHashes)
+          (toReturn, toProcess) <- allowedForProcessing(ordinal, possibleCandidates).map(_.partitionMap(identity))
+          toAdd = selectStateChannels(lastHash, toProcess.flatten)
+        } yield (toAdd, toReturn.flatten.map(_.output))
+
+        private def onlyPossibleReferences(lastHashReferencs: Hash, outputs: List[StateChannelOutputWithHashes]) =
+          outputs
+            .filter(o => (lastHashReferencs :: outputs.map(_.hash)).contains(o.output.snapshotBinary.value.lastSnapshotHash))
+
+        private def allowedForProcessing(ordinal: SnapshotOrdinal, withHashes: List[StateChannelOutputWithHashes]) =
+          ordinalDelay.traverse { delay =>
+            withHashes.groupBy(o => (o.output.address, o.output.snapshotBinary.lastSnapshotHash)).toList.traverse {
+              case (key, outputs) =>
+                firstSeenOrdinalR(key).modify {
+                  case None                                                             => (ordinal.some, Left(outputs))
+                  case Some(firstOrdinal) if firstOrdinal.value + delay > ordinal.value => (firstOrdinal.some, Left(outputs))
+                  case _                                                                => (None, Right(outputs))
+                }
+            }
+          }.map(_.getOrElse(List(Right(withHashes))))
+
+        private def selectStateChannels(lastHash: Hash, stateChannels: List[StateChannelOutputWithHashes]) = {
+          val lastHashForStateChannel = stateChannels.groupByNec(_.output.snapshotBinary.lastSnapshotHash)
+
+          def unfold(lastHash: Hash): Eval[List[StateChannelOutputWithHashes]] =
+            lastHashForStateChannel
+              .get(lastHash)
+              .map(pickMajority)
+              .map { go =>
+                for {
+                  head <- Eval.now(go)
+                  tail <- unfold(go.hash)
+                } yield head :: tail
+              }
+              .getOrElse(Eval.now(List.empty))
+
+          unfold(lastHash).value.toNel.map(_.map(_.output.snapshotBinary).reverse)
+        }
+
+        private def pickMajority(outputs: NonEmptyChain[StateChannelOutputWithHashes]) =
+          outputs.tail
+            .foldLeft(NonEmptyChain(outputs.head)) {
+              case (acc, o) if acc.head.output.snapshotBinary.proofs.length < o.output.snapshotBinary.proofs.length  => NonEmptyChain(o)
+              case (acc, o) if acc.head.output.snapshotBinary.proofs.length == o.output.snapshotBinary.proofs.length => acc.append(o)
+              case (acc, _)                                                                                          => acc
+            }
+            .groupBy(_.hash)
```

Comment:
Does this assume that we may have multiple proposals with the same number of signatures and the same hash and in the next step we are choosing the hash which gets repeated the most times?